hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/prehandle/CompoundSignatureVerificationFuture.java

```diff
@@ -0,0 +1,218 @@
+        canceled = true;
+        return true;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public boolean isCancelled() {
+        return canceled;
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * Due to the asynchronous nature of {@link TransactionSignature}, it is possible that at the time this method
+     * is called, one or more {@link TransactionSignature}s do not yet have a {@link Future}. If that is the case,
+     * then we are clearly not done!
+     */
+    @Override
+    public boolean isDone() {
+        if (canceled) {
+            return true;
+        }
+
+        for (final var future : futures) {
+            if (!future.isDone()) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * Blocks on each {@link Future}. Due to the asynchronous nature of {@link TransactionSignature}, it is possible
+     * that a {@link Future} has not been assigned to one or more {@link TransactionSignature}s. In that case, we will
+     * wait for the {@link Future} before proceeding.
+     */
+    @Override
+    public SignatureVerification get() throws InterruptedException, ExecutionException {
+        for (final var future : futures) {
+            final var verification = future.get();
+            if (!verification.passed()) {
+                return new SignatureVerificationImpl(key, hollowAccount, false);
+            }
+        }
+
+        return new SignatureVerificationImpl(key, hollowAccount, true);
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * Blocks on each {@link Future}. Due to the asynchronous nature of {@link TransactionSignature}, it is possible
+     * that a {@link Future} has not been assigned to one or more {@link TransactionSignature}s. In that case, we will
+     * wait for the {@link Future} before proceeding (up to the timeout).
+     */
+    @Override
+    public SignatureVerification get(final long timeout, final TimeUnit unit)
+            throws InterruptedException, ExecutionException, TimeoutException {
+        var millisRemaining = unit.toMillis(timeout);
+        int failCount = 0;
+        for (final var future : futures) {
+            // We know the maximum number of millis we can wait, so we block for at most that long. If it takes
+            // longer, the future we delegate to here will throw. If it takes less, we recompute how much is
+            // remaining on the next iteration.
+            final var beforeWaiting = System.currentTimeMillis();
+            final var verification = future.get(millisRemaining, TimeUnit.MILLISECONDS);
+            if (!verification.passed() && (failCount++ >= numCanFail)) {
+                return new SignatureVerificationImpl(key, hollowAccount, false);
+            }
+            // Figure out how many milliseconds we still have remaining before timing out
+            millisRemaining -= System.currentTimeMillis() - beforeWaiting;
```

Comment:
I would add a check for `millisRemaining <= 0` and throw a `TimeoutException` in that case.

Reply:
Using @jsync-swirlds's approach to the other variant of this, what if I had this implementation:

```
    @Override
    public SignatureVerification get(final long timeout, final TimeUnit unit)
            throws InterruptedException, ExecutionException, TimeoutException {
        int failCount = 0;
        final var deadline = System.currentTimeMillis() + unit.toMillis(timeout);
        for (final var future : futures) {
            // We know the maximum number of millis we can wait, so we block for at most that long. If it takes
            // longer, the future we delegate to here will throw. If it takes less, we recompute how much is
            // remaining on the next iteration.
            final var now = System.currentTimeMillis();
            final var verification = future.get(deadline - now, TimeUnit.MILLISECONDS);
            if (!verification.passed() && (failCount++ >= numCanFail)) {
                return new SignatureVerificationImpl(key, hollowAccount, false);
            }
        }

        return new SignatureVerificationImpl(key, hollowAccount, true);
    }
```

This gets rid of that line entirely.